google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonToProtoMessage.java

```diff
@@ -0,0 +1,328 @@
+                String.format(
+                    "JSONObject has fields unknown to BigQuery: %s. Set allowUnknownFields to True to allow unknown fields.",
+                    currentScope));
+          } else {
+            continue;
+          }
+        }
+        matchedFields++;
+        if (!field.isRepeated()) {
+          fillField(protoMsg, field, json, jsonName, currentScope, allowUnknownFields);
+        } else {
+          fillRepeatedField(protoMsg, field, json, jsonName, currentScope, allowUnknownFields);
+        }
+      }
+    }
+    if (matchedFields == 0 && topLevel) {
+      throw new IllegalArgumentException(
+          "There are no matching fields found for the JSONObject and the protocol buffer descriptor.");
+    }
+    DynamicMessage msg;
+    try {
+      msg = protoMsg.build();
+    } catch (UninitializedMessageException e) {
+      String errorMsg = e.getMessage();
+      int idxOfColon = errorMsg.indexOf(":");
+      String missingFieldName = errorMsg.substring(idxOfColon + 2);
+      throw new IllegalArgumentException(
+          String.format(
+              "JSONObject does not have the required field %s.%s.", jsonScope, missingFieldName));
+    }
+    return msg;
+  }
+
+  /**
+   * Fills a non-repetaed protoField with the json data.
+   *
+   * @param protoMsg The protocol buffer message being constructed
+   * @param fieldDescriptor
+   * @param json
+   * @param actualJsonKeyName Actual key name in JSONObject instead of lowercased version
+   * @param currentScope Debugging purposes
+   * @param allowUnknownFields Ignores unknown JSON fields.
+   * @throws IllegalArgumentException when JSON data is not compatible with proto descriptor.
+   */
+  private static void fillField(
+      DynamicMessage.Builder protoMsg,
+      FieldDescriptor fieldDescriptor,
+      JSONObject json,
+      String actualJsonKeyName,
+      String currentScope,
+      boolean allowUnknownFields)
+      throws IllegalArgumentException {
+    java.lang.Object val;
+    try {
+      switch (fieldDescriptor.getType()) {
+        case BOOL:
+          protoMsg.setField(fieldDescriptor, new Boolean(json.getBoolean(actualJsonKeyName)));
+          break;
+        case BYTES:
+          protoMsg.setField(fieldDescriptor, json.getString(actualJsonKeyName).getBytes());
+          break;
+        case INT64:
+          val = json.get(actualJsonKeyName);
+          if (val instanceof Integer) {
+            protoMsg.setField(fieldDescriptor, new Long((Integer) val));
+          } else if (val instanceof Long) {
+            protoMsg.setField(fieldDescriptor, new Long((Long) val));
+          } else {
+            throw new JSONException("");
+          }
+          break;
+        case INT32:
+          val = json.get(actualJsonKeyName);
+          if (val instanceof Integer) {
+            protoMsg.setField(fieldDescriptor, new Integer((Integer) val));
+          } else {
+            throw new JSONException("");
```

Comment:
We need specific error info here, this is a field type mismatch, ideally it would be expected Integer but actual is blah.

Reply:
So I threw this empty JsonException error and caught it in the outer loop to throw the actual error which is "JSONObject does not have a <fieldtype> field at <scope>."  If I were to fill in the exception there it would be the same message "JSONObject does not have a <fieldtype> field at <scope>."